22-malloc/libc/string.c

```diff
@@ -25,17 +25,13 @@ void hex_to_ascii(int n, char str[]) {
 
     s32 tmp;
     int i;
-    for (i = 28; i > 0; i -= 4) {
+    for (i = 28; i >= 0; i -= 4) {
```

Comment:
Adding >= here will cause the string to become "0x11000**0**" instead of "0x11000".

Reply:
I don’t think so. You can compile and run the following program to verify:

```c
#include <stdio.h>

int strlen1(char s[]) {
    int i = 0;
    while (s[i] != '\0') ++i;
    return i;
}

void append(char s[], char n) {
    int len = strlen1(s);
    s[len] = n;
    s[len+1] = '\0';
}

void hex_to_ascii(int n, char str[]) {
    append(str, '0');
    append(str, 'x');
    char zeros = 0;

    int tmp;
    int i;
    for (i = 28; i >= 0; i -= 4) {
        tmp = (n >> i) & 0xF;
        if (tmp == 0 && zeros == 0) continue;
        zeros = 1;
        if (tmp >= 0xA) append(str, tmp - 0xA + 'a');
        else append(str, tmp + '0');
    }
}


int main() {
    char s[100];
    hex_to_ascii(0x11000, s);
    printf("%s\n", s);
}
```

Reply:
Yup, you're right. Nice example. I was fairly tired when I did this and just noticed that I didn't remove the extra appends after the loop, which is why I was getting an extra 0 with >=.